File: app/parser.py
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation
import re


MONEY_RE = re.compile(r"(?P<sign>-)?(?:[$€£]\s*)?(?P<amount>\d+(?:,\d{3})*(?:\.\d{1,2})?|\d+)(?:\s*(?P<currency>[A-Za-z]{3}))?")
KIND_WORDS = {"expense", "income", "investment", "transfer"}
ACCOUNT_HINTS = {"cash", "debit card", "credit card", "bank", "checking", "savings", "brokerage"}


@dataclass(frozen=True)
class ParsedTransaction:
    category: str
    account: str
    description: str
    amount: Decimal
    currency: str = "USD"
    kind: str = "expense"


def parse_transaction_message(text: str) -> ParsedTransaction:
    parts = [part.strip() for part in text.split(",") if part.strip()]
    if len(parts) < 3:
        raise ValueError("Use: category, account, description, $amount")

    amount_part = parts[-1]
    amount, currency = _parse_money(amount_part)
    body = parts[:-1]

    kind = "expense"
    if body[0].lower() in KIND_WORDS:
        kind = body.pop(0).lower()

    if len(body) < 2:
        raise ValueError("Use: category, account, description, $amount")

    category = body[0].lower()
    account = body[1].lower()
    description = ", ".join(body[2:]).strip() or category

    if kind == "income":
        amount = abs(amount)
    elif kind in {"expense", "investment"}:
        amount = -abs(amount)

    return ParsedTransaction(
        category=category,
        account=account,
        description=description,
        amount=amount,
        currency=currency,
        kind=kind,
    )


def _parse_money(text: str) -> tuple[Decimal, str]:
    match = MONEY_RE.search(text.strip())
    if not match:
        raise ValueError("Could not find an amount like $4.20")

    try:
        amount = Decimal(match.group("amount").replace(",", ""))
    except InvalidOperation as exc:
        raise ValueError("Amount is not valid") from exc

    if match.group("sign"):
        amount = -amount
    return amount.quantize(Decimal("0.01")), (match.group("currency") or "USD").upper()

File: lib/parser.ts
export type ParsedTransaction = {
  kind: "expense" | "income" | "investment" | "transfer";
  category: string;
  account: string;
  description: string;
  amountCents: number;
};

export type ConciseTransaction = {
  kind: "expense";
  description: string;
  amountCents: number;
};

import { normalizeIdentity } from "@/lib/identity";

const kindWords = new Set(["expense", "income", "investment", "transfer"]);
const moneyPattern = /^(?<sign>-)?(?<amount>\d+(?:,\d{3})*(?:\.\d{1,2})?|\d+)$/;

export function parseTransactionMessage(text: string): ParsedTransaction {
  const parts = text.split(",").map((part) => part.trim()).filter(Boolean);
  if (parts.length < 3) throw new Error("Use: category, account, description, amount");

  const { amountCents: rawAmountCents } = parseMoney(parts.at(-1) || "");
  const body = parts.slice(0, -1);
  let kind: ParsedTransaction["kind"] = "expense";

  if (kindWords.has(body[0]?.toLowerCase())) {
    kind = body.shift()?.toLowerCase() as ParsedTransaction["kind"];
  }

  if (kind === "transfer") {
    throw new Error("Transfers must be created in the dashboard.");
  }

  if (body.length < 2) throw new Error("Use: category, account, description, amount");

  let amountCents = rawAmountCents;
  if (kind === "income") amountCents = Math.abs(amountCents);
  if (kind === "expense" || kind === "investment") amountCents = -Math.abs(amountCents);

  return {
    kind,
    category: normalizeIdentity(body[0]),
    account: normalizeIdentity(body[1]),
    description: body.slice(2).join(", ").trim() || body[0].toLowerCase(),
    amountCents
  };
}

export function parseConciseTransactionMessage(text: string): ConciseTransaction {
  const match = /^\s*(?<money>\d+(?:,\d{3})*(?:\.\d{1,2})?)\s+(?<description>.+?)\s*$/.exec(text);
  if (!match?.groups?.money || !match.groups.description) {
    throw new Error("Use a concise amount and subcategory, such as: 4.20 eat out");
  }
  const { amountCents } = parseMoney(match.groups.money);
  return {
    kind: "expense",
    description: match.groups.description.trim(),
    amountCents: -Math.abs(amountCents)
  };
}

export function isConciseTransactionMessage(text: string) {
  return /^\s*\d+(?:,\d{3})*(?:\.\d{1,2})?\s+\S/.test(text);
}

function parseMoney(text: string) {
  const match = moneyPattern.exec(text);
  if (!match?.groups?.amount) throw new Error("Could not find an amount like 4.20");
  const amount = Number(match.groups.amount.replaceAll(",", ""));
  if (!Number.isFinite(amount)) throw new Error("Amount is not valid");

  const signed = match.groups.sign ? -amount : amount;
  return {
    amountCents: Math.round(signed * 100)
  };
}
